File: 3956-maximum-sum-of-m-non-overlapping-subarrays-i/3956-maximum-sum-of-m-non-overlapping-subarrays-i.cs
using System;
using System.Collections.Generic;

public class Solution
{
    private long UnboundDP(long[] pre, int n, int l, int r)
    {
        long[] dp = new long[n + 1];
        LinkedList<int> dq = new LinkedList<int>();
        long ans = long.MinValue / 2;

        for (int i = l; i <= n; i++)
        {
            int t = i - l;
            long cur = dp[t] - pre[t];

            while (dq.Count > 0 &&
                   dp[dq.Last.Value] - pre[dq.Last.Value] <= cur)
            {
                dq.RemoveLast();
            }

            dq.AddLast(t);

            while (dq.Count > 0 && dq.First.Value < i - r)
            {
                dq.RemoveFirst();
            }

            dp[i] = dp[i - 1];

            if (dq.Count > 0)
            {
                long v = pre[i] + dp[dq.First.Value] - pre[dq.First.Value];
                dp[i] = Math.Max(dp[i], v);
                ans = Math.Max(ans, v);
            }
        }

        return ans;
    }

    public long MaximumSum(int[] nums, int m, int l, int r)
    {
        int n = nums.Length;

        long[] pre = new long[n + 1];
        for (int i = 0; i < n; i++)
        {
            pre[i + 1] = pre[i] + nums[i];
        }

        long neg = long.MinValue / 2;

        if (m >= n / l)
        {
            return UnboundDP(pre, n, l, r);
        }

        long[] prev = new long[n + 1];
        long ans = neg;
        long last = neg;

        for (int k = 1; k <= m; k++)
        {
            long[] cur = new long[n + 1];
            Array.Fill(cur, neg);

            LinkedList<int> dq = new LinkedList<int>();

            for (int i = l; i <= n; i++)
            {
                int t = i - l;
                long val = prev[t] - pre[t];

                while (dq.Count > 0 &&
                       prev[dq.Last.Value] - pre[dq.Last.Value] <= val)
                {
                    dq.RemoveLast();
                }

                dq.AddLast(t);

                while (dq.Count > 0 && dq.First.Value < i - r)
                {
                    dq.RemoveFirst();
                }

                if (dq.Count > 0)
                {
                    cur[i] = pre[i] + prev[dq.First.Value] - pre[dq.First.Value];
                    ans = Math.Max(ans, cur[i]);
                }
            }

            if (ans == last) break;

            last = ans;

            long[] best = new long[n + 1];
            Array.Fill(best, neg);

            for (int i = 1; i <= n; i++)
            {
                best[i] = Math.Max(best[i - 1], cur[i]);
            }

            prev = best;
        }

        return ans;
    }
}

File: 3956-maximum-sum-of-m-non-overlapping-subarrays-i/3956-maximum-sum-of-m-non-overlapping-subarrays-i.dart
import 'dart:math';

class Solution {
  int maximumSum(List<int> nums, int m, int l, int r) {
    int unboundDP(List<int> pre, int n, int l, int r) {
      List<int> dp = List.filled(n + 1, 0);
      List<int> dq = [];
      int head = 0;
      int ans = -(1 << 60);

      for (int i = l; i <= n; i++) {
        int t = i - l;
        int cur = dp[t] - pre[t];

        while (dq.length > head &&
            dp[dq.last] - pre[dq.last] <= cur) {
          dq.removeLast();
        }

        dq.add(t);

        while (dq.length > head && dq[head] < i - r) {
          head++;
        }

        dp[i] = dp[i - 1];

        if (dq.length > head) {
          int v = pre[i] + dp[dq[head]] - pre[dq[head]];
          dp[i] = max(dp[i], v);
          ans = max(ans, v);
        }
      }

      return ans;
    }

    int n = nums.length;

    List<int> pre = List.filled(n + 1, 0);
    for (int i = 0; i < n; i++) {
      pre[i + 1] = pre[i] + nums[i];
    }

    int neg = -(1 << 60);

    if (m >= n ~/ l) {
      return unboundDP(pre, n, l, r);
    }

    List<int> prev = List.filled(n + 1, 0);

    int ans = neg;
    int last = neg;

    for (int k = 1; k <= m; k++) {
      List<int> cur = List.filled(n + 1, neg);

      List<int> dq = [];
      int head = 0;

      for (int i = l; i <= n; i++) {
        int t = i - l;
        int val = prev[t] - pre[t];

        while (dq.length > head &&
            prev[dq.last] - pre[dq.last] <= val) {
          dq.removeLast();
        }

        dq.add(t);

        while (dq.length > head && dq[head] < i - r) {
          head++;
        }

        if (dq.length > head) {
          cur[i] = pre[i] + prev[dq[head]] - pre[dq[head]];
          ans = max(ans, cur[i]);
        }
      }

      if (ans == last) break;

      last = ans;

      List<int> best = List.filled(n + 1, neg);

      for (int i = 1; i <= n; i++) {
        best[i] = max(best[i - 1], cur[i]);
      }

      prev = best;
    }

    return ans;
  }
}